src/HappyEyeBallsConnector.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace React\Socket;
+
+use React\Dns\Model\Message;
+use React\Dns\Resolver\Resolver;
+use React\EventLoop\LoopInterface;
+use React\EventLoop\TimerInterface;
+use React\Promise;
+use React\Promise\CancellablePromiseInterface;
+use InvalidArgumentException;
+use RuntimeException;
+
+final class HappyEyeBallsConnector implements ConnectorInterface
+{
+    private $loop;
+    private $connector;
+    private $resolver;
+
+    public function __construct(LoopInterface $loop, ConnectorInterface $connector, Resolver $resolver)
+    {
+        $this->loop = $loop;
+        $this->connector = $connector;
+        $this->resolver = $resolver;
+    }
+
+    public function connect($uri)
+    {
+
+        if (\strpos($uri, '://') === false) {
+            $parts = \parse_url('tcp://' . $uri);
+            unset($parts['scheme']);
+        } else {
+            $parts = \parse_url($uri);
+        }
+
+        if (!$parts || !isset($parts['host'])) {
+            return Promise\reject(new \InvalidArgumentException('Given URI "' . $uri . '" is invalid'));
+        }
+
+        $host = \trim($parts['host'], '[]');
+        
+        // skip DNS lookup / URI manipulation if this URI already contains an IP
+        if (false !== \filter_var($host, \FILTER_VALIDATE_IP)) {
+            return $this->connector->connect($uri);
+        }
+
+        try {
+            $builder = new HappyEyeBallsConnectionBuilder(
+                $this->loop,
+                $this->connector,
+                $this->resolver,
+                $uri,
+                $host,
+                $parts
+            );
+            return $builder->connect();
+        } catch (\Exception $exception) {
+            return Promise\reject($exception);
+        } catch (\Throwable $throwable) {
+            return Promise\reject($throwable);
+        }
```

Comment:
The try-catch looks like a leftover?

Reply:
Unless an `Error` is throw somewhere down the stack and we won't catch it then on PHP 7.

Reply:
Not sure I follow, why would this piece of code throw an `Error` in the first place?

Reply:
It shouldn't, removed the try catch